### ISSUE TITLE
Feature request: Use indexes for ORDER BY

### ISSUE THREAD

**Issue by rationalism** [neo4j/neo4j#6584]
Currently (in neo4j 2.3.2), this Cypher:

PROFILE MATCH (n:person) RETURN n.email ORDER BY n.email LIMIT 1

starts off with an O(N) NodeByLabelScan, while this Cypher:

PROFILE MATCH (n:person) WHERE n.email < "aaaaaaa@gmail.com" RETURN n.email LIMIT 1

starts with a NodeUniqueIndexSeekByRange (if person:email is indexed), which is much faster for large graphs. I can sort of work around this by saying WHERE n.email < x and then doing binary search on x, but that's clunky and slow. Can ORDER BY be updated to use indexes in 3.0? Thanks.

Relevant StackOverflow thread: http://stackoverflow.com/questions/23835349/does-cyphers-order-by-uses-the-index

**Comment by jonathancrosmer** [comment]
I was surprised this isn't supported---we only discovered it as our data started scaling up.  This is a common use case in SQL databases.  For example
https://www.percona.com/blog/2006/09/01/mysql-order-by-limit-performance-optimization/
MySQL can use multi-column indexes to speed up even a more complex query with WHERE and ORDER BY.  But Neo4j currently doesn't support multi-property indexes, and doesn't even use the single-property index as expected.
For a query like:
`PROFILE MATCH (n:Person) WHERE n.isInteresting = true RETURN n.email ORDER BY n.email LIMIT 1`
I would expect Neo4j to scan the Person(email) index, filter out uninteresting people, and return the first email address.  Depending on the use case, this might be 1 or a few rows searched, or many (if most people are uninteresting).  But as I understand it, Neo4j instead first sorts all Person nodes by email, resulting in O(n log n) complexity regardless of any indexes available.

This author https://dzone.com/articles/neo4j-and-cypher-profiling classifies ORDER BY + LIMIT as an anti-pattern in Neo4j (!) and offers a complex workaround involving simulating an index in the graph model itself.  Unfortunately that is impractical for many cases.

**Comment by jexp** [comment]
@jonathancrosmer how many entries do you have in your Person index?

Would the range scan that @rationalism suggested work for you?

I.e. find a sensible range search upper limit that you can use to subselect a subset of the total entries and then have Neo only filter and sort the subset?

The upper limit could come from the domain or from a one-off query that returns a value that is cached for a longer time (e.g. minutes or hours).

**Comment by jonathancrosmer** [comment]
Thanks for the quick response.
At the moment we have about 500,000.  Queries that do e.g. ORDER BY + LIMIT 100 can result in 5,000,000 database hits.
I think we'll have to do something like that, for now.  It's messy and not 100% functional, but I can't think of a better workaround.

**Comment by jonathancrosmer** [comment]
We are going to do the subselect.  We actually have a property that would almost work reasonably for this purpose, but it can be null (rarely), and we need to include nodes where it is null.  I can't find a way to make Neo4j use the index for this purpose.  For example `WHERE NOT n.x < y` I think excludes nodes where `n IS NULL`, since `null < y` is null, and `NOT NULL` also evaluates to null.  Any kind of `OR` or `COALESCE` etc. seems to prevent the use of an index, and `UNION` can't be combined sensibly with a `WITH` clause.  So I think our only option might be to compute and cache a value like `n.z = COALESCE(n.x, 99999999999)` on every node to be able to use it with the index: `WHERE n.z < y`.

**Comment by jexp** [comment]
You could do a UNION of the query with the range filter + order.
And a second one that uses `WHERE NOT exists(n.foo)` which should also use the index.

**Comment by jonathancrosmer** [comment]
I don't think `WHERE NOT exists(n.foo)` uses the index, based on the result of `PROFILE` and also when I add `USING INDEX`, I get an error: "Cannot use index hint in this context..."

``` cypher
PROFILE MATCH (n:N) USING INDEX n:N(p) WHERE NOT exists(n.p) RETURN n
```

We have another property I can use instead though. And since we don't have this feature yet, https://github.com/neo4j/neo4j/issues/2725 , I'm just doing UNIONing the results of each query and then doing some post-processing in the application layer.
Thanks for looking into this.

**Comment by InverseFalcon** [comment]
Any chance of this being picked up in version 3.1 (or a soon-to-follow patch)? 

We have some use cases looming where we'll need to paginate potentially large sets of data based on indexed properties (names, dates), and the current performance (NodeByLabelScan) won't scale well for large data sets. For simple queries using ORDER BY ... LIMIT 20 on a single unique indexed property, we start breaking into the seconds for processing when we get into the 100's of thousands area. Once we start getting into the millions we're getting into 10's of seconds.

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by jexp** [comment]
You can use APOC for that for the time being, check out:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by InverseFalcon** [comment]
I tried that previously, but it wasn't working for me and I came to the conclusion that this wasn't the purpose of the function. Perhaps I was using it wrong? Given :Users with an int ID property, if I wanted to sort by the ID and get the top 10, I tried something like this:

```
CALL apoc.index.orderedRange("User","ID",null,null,false,10) yield node
return node.ID
```

But the IDs were all over the place, not in any kind of order, and certainly not from what should have been the set of the last 10 IDs. Changing the boolean to true didn't help either (unsure what sort-relevance refers to, the APOC documentation seems sparse on documenting what parameters mean). 

If you can provide an example of how I should be using this, I would greatly appreciate it.

This was with Neo4j 3.0.3 and APOC version 1.0.0 (or 1.1.0, not sure at the moment), so if there was a bug fixed in a later APOC version that would explain it.

**Comment by jexp** [comment]
Can you raise the apoc related issue please on the apoc repo. And only the feature request here. 

Von meinem iPhone gesendet

**Comment by stigrv** [comment]
I too think this would make a great addition to Neo to be able to order by an index.

I have a list of elements I want to order by created date, and have therefore added an index to this property. As most people, I was suprised that adding an index did no change to the sorting performance whatsoever. I've tried to limit the amount of data by limiting the content to what has been created during the last year. This increased the performance somwhat. The problem will be when more data gets added to the database during heavier usage of the system. I could of course limit the timewindow even further, but I don't see that as a solution. The user is still not able to paginate through all the data, which is a very valid usecase in our set of data.

Are these usecases not going to be covered by Neo, or am I approaching this incorrectly? We do have a lot of connected data, so a graph database seems like the correct choice for us. As long as I have an element, further lookup is extremely fast, but looking up the initial entrypoint is problematic, still.

**Comment by zmarty** [comment]
I am trying to not be negative but this is the second indexing feature that I am astounded does not exist in neo4j. If neo4j is ever to be seen as a serious database it needs to implement a bare minimum of SQL-like functionality.

Our use case is the following: retrieve a list of items sorted by a static rank. As we have many items, this is **extremely** inefficient currently.

**Comment by ban1989ban** [comment]
Version 3.2 is almost there but with what I have read in changeset, I dont think this feature is still not implemented. I guess this is very basic feature & it should be there.  Any chance this feature comes in near future? We wont be able to use it without this feature.

**Comment by stigrv** [comment]
It seems the current solution is to integrate against Elasticsearch. Take a look at GraphAware (http://graphaware.com/). They do this integration, and from what I've tested, their solution works quite well.

I guess that if Neo4J implements the feature described in this bug, these guys would be out of business quite soon.

**Comment by zmarty** [comment]
neo4j 3.2 is now officially out, and this basic feature is still MIA

**Comment by Lojjs** [comment]
Hi! We in the Cypher Team are now in a planning phase for the coming Neo4j releases. This is one of the features which are highly prioritized right now so hopefully it will soon be implemented. In the meantime I refer to APOC and @jexp. 

I will come back to you when I have more info. Thanks for you patience!

**Comment by smullins3000** [comment]
Jumping on the bandwagon for this, I was very surprised when I discovered this fact, that "order by" when used with "limit" is not at all scalable (due to not using the index on the sorted properties.
I agree with others that sorting and paging are table stakes for a number of applications including I believe mine.
My application needs to pull together a list of books and all the different ways that a customer might buy some element of that book. There are two primary ways, where a book might have an audio version and might have a number of different "viewable" media (paper, PDF, amazon, etc.). That media must exist and the file for that type must also exist. Finally, there has to be a purchasable "product" associated with the book element (not all things are purchasable!). 
This can be really, really large as there are a lot of books, and as well we want people to be able to sort and browse through the available listing as needed. Hence, "order by" + "limit".

```
match (p:Book)-[:hasaudio|:hasviewable]->(ml)<-[*]-(prod:Product)
where ((p)-[:hasaudio]->(ml:Audio)) 
    OR ((p)-[:hasviewable]->(ml:Viewable {availabilityState:"available"})-[:hasfile]->(:File))
return distinct p.bookid, p.title order by p.title limit 50
```

Adding the "order by" clause in a graph with ~200k nodes and ~200k directed edges causes the query tim to go from <20ms to >1000ms. That increase scales very directly with the size of the graph.

**Comment by pbnjay** [comment]
@Lojjs any word on this? My use case would REALLY benefit! Would be happy to beta-test...

**Comment by Lojjs** [comment]
@pbnjay I don't have any news for now, but this is still a high prioritized feature. I will let you know when I have more details.

**Comment by smullins3000** [comment]
Thank you. I could definitely see beta testing this as well.

**Comment by matthewoates** [comment]
Hi all,

This seems like a massive feature which if done poorly (or perhaps if done at all) could a divergence from the graph paradigm.

Nevertheless, I'd love to hear if there is any momentum on this, or if there's any way we can help out. (planning, contributing, testing, etc) We're a startup with 8 engineers and would like Neo4J to power our real time recommendations as well as application data store. Without indexes such as this, we'd also need to set up something like ElasticSearch to have feature parity.

cc @Lojjs

**Comment by matthewoates** [comment]
Does this PR cover this requested feature?

https://github.com/neo4j/neo4j/pull/10702

**Comment by InverseFalcon** [comment]
@matthewoates No, that looks like an unrelated PR. 

I believe work on this feature is being tracked internally, there may not be a corresponding PR.

**Comment by bonham000** [comment]
Would love to see this feature!

**Comment by prathle** [comment]
tl;dr: This is an obvious and necessary feature. It's been prioritized and about 50% of the work has been done. It will land in Neo4j 3.5, which is targeted for EOY '18. 

Let me explain below why it's taken as long as it has:

We've been in the midst of a major forklift of our indexing implementation over the last three releases (Neo4j 3.2, 3.3, and 3.4). We're doing this to address what is _by far_ the biggest write bottleneck in Neo4j, namely the use of Lucene for schema indexes. Lucene throughput is great if you use it in an eventually-consistent manner. We cannot however (at least not without subjecting users to divergence between the graph & index, which we consider corruption) which is why we've had to bite the bullet and create our own indexing implementation. 

The new index is designed specially for Neo4j, and its main purpose is to support high levels of parallelism & scaling, with ACID writes. Micro-benchmarks vs. Lucene show that it's 5-10x better in terms of throughput for writes. In Neo4j 3.2 it was introduced for the label scan index, then for numerics in Neo4j 3.3, and finally for strings in Neo4j 3.4. It's been a long time coming and has been our main focus for indexes.

Some good news is that we recently completed the kernel work necessary to support index-backed order by with the new indexes for numerics. The Cypher work still needs to get done, and is scheduled for the next release; and at that time we'll do the kernel work to support index-backed order by for strings. For those of you who rely on more complex search functionality with strings: fear not, we'll have you covered in the same timeframe as this improvement.

We'll post updates to this thread as we get closer and there's something to test. Thanks all for your forbearance.

The workaround mentioned above by @jexp may be helpful in the meantime as a stopgap, in at least some cases:

https://neo4j-contrib.github.io/neo4j-apoc-procedures/#_schema_index_queries

**Comment by Tin-Nguyen** [comment]
@prathle Do we have any workaround for `ORDER BY` performance with indexing for `Neo4j 2.3.8`, as I know the version doesn't support `APOC` yet?

**Comment by prathle** [comment]
@Tin-Nguyen you could try doing the same thing with an unmanaged extension. 

That said, Neo4j 2.3 is > 2.5 years old. Neo4j 3.3 has, without exaggeration, hundreds of engineer-years worth of improvements over Neo4j 2.3. It is, in many respects a significantly different product. 

As time goes on, velocity is accelerating. Neo4j increased the size of our full-time product engineering staff by 50% last year. The result is a Neo4j 3.4 release that will be very substantial across the feature, performance, scalability, and operability dimensions. 

All that to say that this is probably a good time to start planning your upgrade. New features are scheduled into x.0 and x.y releases; whereas the 2.3.x series is, by definition, frozen in time.

**Comment by ericsantosn** [comment]
Hey guys, what's the status of this feature, is there any estimation of when it will be released? I was looking for more information in the project but I couldn't find anything for it.

**Comment by ericsantosn** [comment]
Reading the thread I assume there's no timeline for this feature right? my case is just same as all guys, very poor performance when I sort by attribute on neo4j and we   it gets worst and worst, we're already at

**Comment by sherfert** [comment]
The first version of an index-backed-order-by is available in the latest `3.5.0-beta01` . Please check the [release notes](https://github.com/neo4j/neo4j/wiki/Neo4j-3.5-changelog#350-beta01) to find out what queries are supported.

**Comment by methyl** [comment]
As 3.5.0 is now released and I can't really find it in the changelog, is ORDER BY now using indexes properly without limitations such as only ASC order being supported?

**Comment by Lojjs** [comment]
@methyl DESC order is supported but all queries are not supported yet, we are still working on improvements to this feature. To see what is supported in 3.5.0, please read the documentation here: https://neo4j.com/docs/cypher-manual/current/query-tuning/cypher-index-values-order/

**Comment by methyl** [comment]
I'm playing with it and it seems it doesn't handle my usecase:
```
PROFILE
MATCH (p:Person {name: "Tom Hanks"})-[:ACTED_IN]->(m:Movie)
WHERE m.box_office_earnings > 0
RETURN m.movie
ORDER BY m.box_office_earnings
LIMIT 10
```
In my case, there can be hundreds of thousands of movies for some actors so full scan is not reasonable, that's why I really count on having this feature in place, otherwise I have to resort to SQL database to get it done.

Is this something you will support eventually?

This uses indexes just fine:
```
PROFILE
MATCH (m:Movie)
WHERE m.box_office_earnings > 0
RETURN m.movie
ORDER BY m.box_office_earnings
LIMIT 10
```

**Comment by sherfert** [comment]
@methyl 
Does it work if you add `USING INDEX m:Movie(box_office_earnings)` to the first query?

Broader answer: Yes, we want to support such queries to choose the cheaper index automatically because of ordering in the future, but it didn't make it into 3.5. Right now it will (unless you provide hints) solve this part of the query `(p:Person {name: "Tom Hanks"})-[:ACTED_IN]->(m:Movie) WHERE m.box_office_earnings > 0` the cheapest way possible (which is by starting at `p`) and then realize that this means it has to sort results.

**Comment by methyl** [comment]
@sherfert when I add using index, it seems to stop using index for `p:Person`, so that it still takes forever. Adding two using index:
```
using index p:Person(name)
using index m:Movie(box_office_earnings)
```

does not seem to solve issue either, it won't load even if I don't order anymore.

**Comment by sherfert** [comment]
To use both indexes, the planner needs to realize that the cheapest plan that does not require sorting, is a HashJoin with `m` on the right-hand-side. That, unfortunately, you will not get to work, even with hints, in 3.5. Hopefully 4.0 will have that capability.

**Comment by boggle** [comment]
@craigtaverner @sherfert Any update on this in light of recent changes?

**Comment by Tin-Nguyen** [comment]
hi @sherfert , I'm back to this discussion thread here, as we upgrade to use the latest version of 3.5 (3.5.14). I'm wondering if you have any documentation related to how to use new indexing to speed up the query?

**Comment by sherfert** [comment]
Here is the documentation @Tin-Nguyen https://neo4j.com/docs/cypher-manual/current/query-tuning/advanced-example/#advanced-query-tuning-example-index-backed-order-by

Also, the broad scope of this feature is done in 3.5 and even more in 4.0. I am closing this issue. If anyone finds a query that should use index ordering, but doesn't, please open a new issue.
